add: mux overview (#270)
libp2p/docs: content/concepts/multiplex/overview.md

## Overview

Stream multiplexing (or stream muxing) is a method to send multiple streams
of data over a single communication link. It combines multiple signals into one

> unified signal. The unifed signal can then be demultiplexed (or demuxed), where

Reviewer: This was a stale review that got pushed. marking as resolved

> the output becomes available in an application. This allows

Reviewer 2: I'd prefer to not talk about signals here. Let's keep this to streams and connections.

> Stream multiplexing is done to share a single TCP connection using unique port numbers

Reviewer 2: I suggest a paragraph:
* explaining that establishing a libp2p connection can be expensive and take a few round-trips, e.g. when hole punching is involved
* we only want to do this once, then reuse that connection
* opening a stream is very low-overhead, and doesn't impose any latency penalty

> that distinguish streams between multiple processes (such as kademlia and gossipsub)

Reviewer: This was a stale review that got pushed. marking as resolved

> In particular, stream muxing is done to share a single connection using unique port numbers

Reviewer 2: Port numbers are not really relevant here. We actually reuse port numbers in some transports, so they're not really unique.

> 

Reviewer: This was a stale review that got pushed. marking as resolved

> 

Reviewer: feel free to incorporate parts of this into https://github.com/libp2p/docs/pull/270#discussion_r1095080174

## Muxers in libp2p

> Stream muxers are pluggable in libp2p. A given libp2p application can support several

Reviewer 2: ```suggestion
Stream muxers are pluggable in libp2p. A given libp2p host can support several
```

> 

Reviewer 2: We should mention here that this is also used outside of libp2p. For example, HTTP/2 introduced streams into HTTP, allowing issuing many HTTP requests in parallel on the same connection.

> Establishing a libp2p connection can be expensive and take a few round-trips. Peers

Reviewer 2: Mention that this especially applies when doing NAT traversal / hole punching, and link to the respective document.

> muxers simultaneously and choose one as the default, and allowing an application to
> fall back to a widely-supported multiplexer if a remote peer does not support a preferred
> choice.

Reviewer 2: ```suggestion
muxers simultaneously. The stream muxer is negotiated between the two nodes during the handshake.
Having such a negotiation protocol in place allows libp2p to adopt new muxers in the future, while at
the same time keeping backwards-compatibility with currently deployed muxers.
```

> should aim to establish a connection once, then reuse that connection by opening a stream,
> which has a low overhead and does not impose any latency penalty.

Reviewer 2: ```suggestion
therefore aim to establish a connection once, then run many streams over the same connection. Opening
has a low resource overhead and does not impose any latency penalty.
```

> Stream muxers are pluggable in libp2p. A given libp2p host can support several

Reviewer 2: Mention that some protocols have streams natively, and that we only need pluggable muxers for protocols that don't.

Author: mentioned in the alert

allows peers to initialize and use the same [transport](../../transports/overview)

> connection across the lifetime of their interaction with another peer.

Reviewer 2: I think we explained that above.

{{< alert icon="💡" context="info">}}

> libp2p's multiplexing happens at the application layer, meaning the

Reviewer 2: That's only true for TCP, for QUIC and some other transports, stream multiplexing happens at the transport layer.

## Overview

Stream multiplexing (or stream muxing) is a method to send multiple streams of
data over a single communication link. It allows peers to use the same connection
across the lifetime of their interaction with another peer.

In particular, stream muxing is done to share a single connection between multiple
protocols used by applications to make connection and transmission more efficient.
This is especially useful when doing NAT traversal or hole punching,

> as it allows for efficient use of limited resources. This is also used outside of

Reviewer 2: Add this from the next section:
```
Opening and accepting a new stream has a low resource overhead and does not impose any
latency penalty.
``

libp2p, for example, HTTP/2 introduced streams into HTTP, allowing for many HTTP

> requests in parallel on the same connection.

Reviewer: I agree. This document should also be prefaced with a paragraph on why stream multiplexing is necessary. It hints at it but doesn't explicitly mention why it's critical.

I would add:

libp2p is built on top of a stream abstraction (link to the protocols page) and uses a bi-directional message stream to send data between peers.
However, simply having a single message stream over a connection between two peers is not good enough.
This limits the scalability of applications and in fact creates a bottleneck.
This is because each peer on the other side of the connection could be running multiple applications, each sending and waiting for messages over the single stream.

With only a single stream to use, application A would have to wait for application B's use of the connection to end before another A could start sending and receiving messages over the lone stream.

Therefore, to eliminate this bottleneck and allow more applications to talk to each other over the single connection, libp2p employs stream multiplexing.
This allows a libp2p node to create many virtual connections inside a single connection and allows a node to send multiple different streams of messages over those different virtual connections.

> requests in parallel on the same connection.

Author: It was previously suggested to avoid the mention of connections. Agreed though, this is important.

## Muxers in libp2p

Establishing a libp2p connection can be expensive and take a few round-trips.
Peers therefore aim to establish a connection once, then run many streams over
the same connection. Opening has a low resource overhead and does not impose any

> latency penalty.

Reviewer 2: This is kind of repeating what we had in the section before. I suggest to move the sentence about stream opening to that section.

> 

Reviewer 2: Can we add a paragraph here that says that some protocols come with native streams (QUIC, WebTransport, WebRTC), and libp2p doesn't need to do anything in that case.

Reviewer: agree

Stream muxers are pluggable in libp2p. A given libp2p host can support several
muxers simultaneously. The stream muxer is negotiated between the two nodes during
the handshake. Having such a negotiation protocol in place allows libp2p to adopt

> new muxers in the future while simultaneously keeping backward compatibility with

Reviewer 2: ```suggestion
new muxers in the future, while simultaneously keeping backward compatibility with
```

currently deployed muxers.

{{< alert icon="💡" context="info">}}

> Developers writing libp2p applications rarely need to interact with stream
> multiplexers directly, except during initial configuration to control which modules
> are enabled. Some protocols have streams natively, and only pluggable muxers are
> needed for protocols that don't.

Reviewer 2: ```suggestion
Developers writing libp2p applications rarely need to interact with stream
multiplexers directly, except during initial configuration. A libp2p connection always
is a stream-multiplexed connection, and the libp2p stack takes care of negotiating and
setting up a stream multiplexer.
```

## Overview

> Network protocols often use bi-directional message streams built on top of a stream
> abstraction to transmit data between nodes. However, relying on a single message

Reviewer: ```suggestion
libp2p is built on top of a stream abstraction and uses a bi-directional message stream to send data between peers.
However, relying on a single message
```

Author: The idea is to introduce stream multiplexing "in general" and then relate it to libp2p

Reviewer: Understood, feel free to ignore

> stream can result in scalability issues and bottlenecks, as one application would

Reviewer: ```suggestion
stream over a connection between two peers can result in scalability issues and bottlenecks.
Each peer on either side of the connection may run multiple applications sending and waiting for data over the stream.
A single stream would block applications on one another, as one application would
```

need to wait for another to finish utilizing the stream before being able to send
and receive its own messages.

> To overcome this issue, network connections employ stream multiplexing, which allows

Reviewer: ```suggestion
To overcome this issue, libp2p enables applications to employ stream multiplexing.
Multiplexing allows
```

> for the creation of multiple virtual connections within a single connection. This

Reviewer: ```suggestion
for the creation of multiple "virtual" connections within a single connection. This
```

enables nodes to send multiple streams of messages over separate virtual connections,
providing a scalable solution that eliminates the bottleneck created by a

> single stream. An example is when HTTP/2 introduced streams into HTTP,

Reviewer: ```suggestion
single stream. 
Two libp2p peers may have a single TCP connection and use different port numbers to distinguish streams.
Then different applications/processes like Kademlia or GossipSub used by an application like IPFS would get their own stream of data and make transmission more efficient.
Stream multiplexing makes it so that applications or protocols running on top of libp2p think that they’re the only ones running on that connection.
Another example is when HTTP/2 introduced streams into HTTP,
```

Author: We can add an example, but to the previous point, I'd prefer introducing stream muxing in a way that is generic (not specific to libp2p), mention the case with HTTP/2, and then lead into libp2p and an example with libp2p.

Reviewer: Understood, feel free to ignore

allowing for many HTTP requests in parallel on the same connection.

> In libp2p, stream muxing is used to share a single connection between various protocols,

Reviewer: ```suggestion
In summary, stream muxing can be used by applications on top of libp2p to share a single connection between various protocols,
```
